rfcs/text/0005_route_handler.md

```diff
@@ -0,0 +1,184 @@
+  payload: T,
+  options?: HttpResponseOptions
+) => KibanaResponse<T>;
+
+const kibanaResponseToolkit = {
+  ok: <T extends HttpResponsePayload>(payload: T, options?: HttpResponseOptions) =>
+    new KibanaResponse(200, payload, options),
+  accepted: <T extends HttpResponsePayload>(payload: T, options?: HttpResponseOptions) =>
+    new KibanaResponse(202, payload, options),
+  noContent: (options?: HttpResponseOptions) => new KibanaResponse(204, undefined, options)
+```
+
+##### Redirection
+The server wants a user to perform additional actions.
+```typescript
+const kibanaResponseToolkit = {
+  redirected: (url: string, options?: HttpResponseOptions) => new KibanaResponse(302, url, options),
+  notModified: (options?: HttpResponseOptions) => new KibanaResponse(304, undefined, options),
+```
+
+#####  Error
+Server signals that request cannot be handled and explains details of the error situation
+```typescript
+// Supports attaching additional data to send to the client
+interface ResponseError extends Error {
+  meta: {
+    data?: JSONValue;
+    code?: string; // error code to simplify search, translations in i18n, etc.
+    link?: string; // link to the docs
+  }
+}
+
+const kibanaResponseToolkit = {
+  // Client errors
+  badRequest: <T extends ResponseError>(err: T, options?: HttpResponseOptions) =>
+    new KibanaResponse(400, err, options),
+  unauthorized: <T extends ResponseError>(err: T, options?: HttpResponseOptions) =>
+    new KibanaResponse(401, err, options),
+
+  forbidden: <T extends ResponseError>(err: T, options?: HttpResponseOptions) =>
+    new KibanaResponse(403, err, options),
+  notFound: <T extends ResponseError>(err: T, options?: HttpResponseOptions) =>
+    new KibanaResponse(404, err, options),
+  conflict: <T extends ResponseError>(err: T, options?: HttpResponseOptions) =>
+    new KibanaResponse(409, err, options),
+
+  // Server errors
+  internal: <T extends ResponseError>(err: T, options?: HttpResponseOptions) =>
+    new KibanaResponse(500, err, options),
+```
+
+##### Generic
+If a custom respond is required
+```typescript
+interface GenericOptions extends HttpResponseOptions {
+  statusCode: StatusCode;
+}
+export const kibanaResponseToolkit = {
+  generic: <T extends HttpResponsePayload>(payload: T, {statusCode, ...options}: GenericOptions) =>
+    new KibanaResponse(statusCode, payload, options),
+```
+# Drawbacks
+- `Handler` is not compatible with Legacy platform implementation when anything can be returned or thrown from handler function and server send it as a valid result. Transition to the new format may require additional work in plugins.
+- `Handler` doesn't cover **all** functionality of Legacy server at the current moment. For example, we cannot render anything in New platform yet and in this case, we have to proxy the request to the Legacy platform endpoint to perform rendering. All such cases should be considered in an individual order.
```

Comment:
Render? I'm not sure what this point is referring too exactly.

Reply:
Handler cannot render a view and users have to relay on Legacy platform to perform `renderApp` / `renderAppWithDefaultConfig` call.

Reply:
Got it. These are special cases right now (login page for example). I agree these should be separate concerns from this RFC.